test/gtest/uct/test_md.cc

```diff
@@ -359,6 +359,8 @@ UCS_TEST_P(test_md, mem_type_detect_mds) {
         UCS_TEST_MESSAGE << ucs_memory_type_names[alloc_mem_type] << ": "
                          << "sys_dev[" << static_cast<int>(mem_attr.sys_dev)
                          << "] (" << dev_name << ")";
+
+        free_memory(address, static_cast<ucs_memory_type_t>(alloc_mem_type));
```

Comment:
Leads to errors like:
```
[ RUN      ] mem_type/test_memtype_cache.basic/2 <2>
/hpc/newhome/yosefe/workspace/hpc/ucx/contrib/../test/gtest/ucs/test_memtype_cache.cc:47: Failure
Expected equality of these values:
  UCS_MEMORY_TYPE_UNKNOWN
    Which is: 5
  mem_info.type
    Which is: 2
cuda-managed
/hpc/newhome/yosefe/workspace/hpc/ucx/contrib/../test/gtest/ucs/test_memtype_cache.cc:47: Failure
Expected equality of these values:
  UCS_MEMORY_TYPE_UNKNOWN
    Which is: 5
  mem_info.type
    Which is: 2
cuda-managed
[  FAILED  ] mem_type/test_memtype_cache.basic/2, where GetParam() = 2 (0 ms)

```

Reply:
@yosefe in ucm/cuda/cudamem.c looks like we're assigning all freed cuda memory as MEMORY_TYPE_CUDA and we don't assigned freed type as MEMORY_TYPE_CUDA_MANAGED even if memory was allocated through cuMemAllocManaged -- but [cuMemFree](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gb347ded34dc326af404aa02af5388a32) is the same API used to free managed memory as well. 

Code from cudamem.c:
```C
UCM_CUDA_FREE_FUNC(cuMemFree, CUresult, CUdeviceptr, UCS_MEMORY_TYPE_CUDA)
UCM_CUDA_FREE_FUNC(cuMemFree_v2, CUresult, CUdeviceptr, UCS_MEMORY_TYPE_CUDA)

```
Either we have to detect memory type while calling free or handle the test. IMO, we should detect to avoid the call to cuMemGetAddressRange as well (as it isn't supported on managed memory) and then dispatch free event using the right detected memory type.

Reply:
Could be, but memory type is not used in free event. I've verified the fix above really fixes the gtest failure.

Reply:
@yosefe are there comments that need to be addressed in this PR?

Reply:
squashed the commits